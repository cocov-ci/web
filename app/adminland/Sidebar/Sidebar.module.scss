.loading {
  padding: 9px;
}
