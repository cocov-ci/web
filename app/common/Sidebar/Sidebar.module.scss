@use "sass:color";
@use "sass:map";
@import "styles/variables.module";

.base {
  padding: 9px;
}

.item {
  cursor: pointer;
  padding: 6px 13px;
  display: flex;
  min-width: 225px;
  justify-content: space-between;
  align-items: center;
  font-variation-settings: "wght" 500;
  margin-bottom: 3px;
  border-radius: 8px;
  border: 1px solid transparent;
  line-height: 19px;

  .label {
    flex: 1;
    font-size: 15px;
    color: $colorBlack;
    letter-spacing: 0;
    text-transform: capitalize;
  }

  .counter {
    font-size: 12px;
    color: map.get($colorSidebar, "counter");
    letter-spacing: 0.96px;
    text-align: right;
  }

  &.selected {
    cursor: default;
    background: map.get($colorListItemActive, "background");
    border: 1px solid map.get($colorListItemActive, "border");
  }

  &:not(.selected):hover {
    background: color.adjust(map.get($colorListItemActive, "background"), $alpha: -0.5);
    border: 1px solid color.adjust(map.get($colorListItemActive, "border"), $alpha: -0.5);
  }
}
