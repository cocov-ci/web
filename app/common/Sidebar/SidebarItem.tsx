'use client'

import classNames from 'classnames'
import React from 'react'

import styles from './Sidebar.module.scss'

import { Item } from './index'

interface SidebarItemProps {
  item: Item
  selected: boolean
  onClick?: (item: Item) => void
}

export const SidebarItem = ({ item, selected, onClick }: SidebarItemProps) => {
  return (
    <a
      className={classNames(styles.item, {
        [styles.selected]: selected,
      })}
      href="javascript:void(0)"
      onClick={() => onClick && onClick(item)}
    >
      <span className={styles.label}>{item.name}</span>
      {item.counter !== undefined && (
        <span className={styles.counter}>{item.counter}</span>
      )}
    </a>
  )
}
