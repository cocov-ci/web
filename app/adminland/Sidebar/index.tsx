'use client'

import { useRouter } from 'next/navigation'
import { useMemo } from 'react'

import Loading from 'app/common/Loading'
import Sidebar from 'app/common/Sidebar'

import styles from './Sidebar.module.scss'

interface SidebarProps {
  loading: boolean
  currentSelectedPath: string | null
}

const SidebarComponent = ({ loading, currentSelectedPath }: SidebarProps) => {
  const router = useRouter()

  const data = useMemo(
    () => [
      {
        id: 0,
        name: 'General',
        href: '',
      },
      {
        id: 1,
        name: 'Service Tokens',
        counter: 0,
        href: '/service-tokens',
      },
      {
        id: 2,
        name: 'Secrets',
        counter: 0,
        href: '/secrets',
      },
      {
        id: 3,
        name: 'Repositories',
        counter: 0,
        href: '/repositories',
      },
      {
        id: 4,
        name: 'Users',
        counter: 0,
        href: '/users',
      },
    ],
    [0],
  )

  if (loading)
    return (
      <Loading
        className={styles.loading}
        count={5}
        height="30px"
        type="skeleton"
        width="230px"
      />
    )

  const onClick = (id: number) => {
    router.push(`/adminland/${data[id].href}`)
  }

  if (!data) return null

  return (
    <Sidebar
      className={styles.sidebar}
      defaultSelectedId={data.find(i => i.href == currentSelectedPath)?.id}
      items={data}
      onSelectItem={item => onClick(parseInt(item.id as string))}
      width="250px"
    />
  )
}

export default SidebarComponent
