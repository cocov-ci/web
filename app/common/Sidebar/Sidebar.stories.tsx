import type { Meta, StoryObj } from '@storybook/react'

import Sidebar from './index'

const meta: Meta<typeof Sidebar> = {
  title: 'Common/Sidebar',
  component: Sidebar,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof Sidebar>

export const Default: Story = {
  args: {
    items: [
      { id: 0, name: 'All Categories', counter: 920 },
      { id: 1, name: 'Complexity', counter: 0 },
      { id: 2, name: 'Convention', counter: 822 },
      { id: 3, name: 'Duplication', counter: 98 },
      { id: 4, name: 'Performance', counter: 0 },
      { id: 5, name: 'Potential Bug', counter: 98 },
      { id: 6, name: 'Security', counter: 0 },
      { id: 7, name: 'Style', counter: 98 },
      { id: 8, name: 'No counter' },
    ],
    defaultSelectedId: 0,
    width: '225px',
  },
}
