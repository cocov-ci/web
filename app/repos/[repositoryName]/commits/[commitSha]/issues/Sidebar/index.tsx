'use client'

import Loading from 'app/common/Loading'
import Sidebar from 'app/common/Sidebar'
import { CommitsSourcesResponseProps } from 'types/Commits'

import styles from './Sidebar.module.scss'

interface SidebarItemsProps {
  id: number
  name: string
  counter: number
}

interface SidebarProps {
  allItemsText: string
  data: CommitsSourcesResponseProps
  loading: boolean
}

const SidebarComponent = ({ allItemsText, data, loading }: SidebarProps) => {
  if (loading)
    return (
      <Loading
        className={styles.loading}
        count={5}
        height="30px"
        type="skeleton"
        width="230px"
      />
    )

  if (!data) return null

  const allItems = {
    id: 0,
    name: allItemsText,
    counter: Object.values(data).reduce((prev, next) => prev + next),
  }

  const items: SidebarItemsProps[] = Object.keys(data)
    .sort((a, b) => a.localeCompare(b))
    .map((item: string, index: number) => ({
      id: index + 1,
      name: item,
      counter: data[item],
    }))

  return (
    <Sidebar
      className={styles.sidebar}
      items={[allItems, ...items]}
      selectedId={0}
      width="250px"
    />
  )
}

export default SidebarComponent
