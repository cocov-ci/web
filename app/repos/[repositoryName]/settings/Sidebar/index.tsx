'use client'

import Loading from 'app/common/Loading'
import Sidebar from 'app/common/Sidebar'

import styles from './Sidebar.module.scss'

interface SidebarItemsProps {
  id: number
  name: string
  counter?: number
}

interface SidebarProps {
  data: SidebarItemsProps[]
  loading: boolean
  defaultSelectedItem: string | null
}

const SidebarComponent = ({
  data,
  loading,
  defaultSelectedItem,
}: SidebarProps) => {
  if (loading)
    return (
      <Loading
        className={styles.loading}
        count={2}
        height="30px"
        type="skeleton"
        width="230px"
      />
    )

  if (!data) return null

  return (
    <Sidebar
      className={styles.sidebar}
      defaultSelectedId={
        data?.filter(item => item.name === defaultSelectedItem)[0]?.id || 0
      }
      items={data}
      onSelectItem={item => null}
      width="250px"
    />
  )
}

export default SidebarComponent
