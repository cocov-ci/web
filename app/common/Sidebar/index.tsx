import classNames from 'classnames'
import React from 'react'

import styles from './Sidebar.module.scss'
import { SidebarItem } from './SidebarItem'

export interface Item {
  id: string | number
  name: string
  counter?: number
}

interface SidebarProps {
  items: Array<Item>
  selectedId: string | number | undefined
  onSelectItem?: (item: Item) => void
  className?: string
  width?: string
}

const Sidebar = ({
  items,
  selectedId,
  onSelectItem,
  className,
  width,
}: SidebarProps) => {
  return (
    <div
      className={classNames(styles.base, className)}
      style={{ width: width }}
    >
      {items.map(item => (
        <SidebarItem
          item={item}
          key={item.id}
          onClick={onSelectItem}
          selected={selectedId === item.id}
        />
      ))}
    </div>
  )
}

export default Sidebar
