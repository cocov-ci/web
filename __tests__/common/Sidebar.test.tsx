import { fireEvent, render, screen } from '@testing-library/react'

import Sidebar from 'app/common/Sidebar'

import '@testing-library/jest-dom'

const items = [
  { id: 0, name: 'All Categories', counter: 920 },
  { id: 1, name: 'Complexity', counter: 130 },
]

describe('common/Sidebar', () => {
  it('renders Sidebar', () => {
    render(<Sidebar defaultSelectedId={0} items={items} />)

    expect(screen.getByText('All Categories')).toBeVisible()
    expect(screen.getByText('920')).toBeVisible()
    expect(screen.getByText('Complexity')).toBeVisible()
    expect(screen.getByText('130')).toBeVisible()
  })

  it('triggers onSelectItem event', () => {
    const handleEvent = jest.fn()
    const { container } = render(
      <Sidebar
        defaultSelectedId={0}
        items={items}
        onSelectItem={handleEvent}
      />,
    )

    expect(container.querySelector('.selected .label')?.textContent).toEqual(
      'All Categories',
    )

    fireEvent.click(screen.getByText('Complexity'))

    expect(container.querySelector('.selected .label')?.textContent).toEqual(
      'Complexity',
    )
    expect(handleEvent).toHaveBeenCalledTimes(1)
  })

  it('renders Sidebar snapshots', () => {
    const { container } = render(
      <Sidebar defaultSelectedId={0} items={items} />,
    )

    expect(container).toMatchSnapshot()
  })
})
