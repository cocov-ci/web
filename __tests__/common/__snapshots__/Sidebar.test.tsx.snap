// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Sidebar renders Sidebar snapshots 1`] = `
<div>
  <div
    class="base"
  >
    <a
      class="item selected"
      href="javascript:void(0)"
    >
      <span
        class="label"
      >
        All Categories
      </span>
      <span
        class="counter"
      >
        920
      </span>
    </a>
    <a
      class="item"
      href="javascript:void(0)"
    >
      <span
        class="label"
      >
        Complexity
      </span>
      <span
        class="counter"
      >
        130
      </span>
    </a>
  </div>
</div>
`;
